### DISCUSSION TITLE
Add a 3rd resource type or a 2nd place to inject resources

### DISCUSSION THREAD

**Discussion by markdav-is** [oqtane/oqtane.framework#776] [Ideas]
We currently have two resources types: Stylesheet & Script.  When listing resource using IHostResources or ModuleBase  Resources property, those resources are injected into the head tag of the main layout in _host.cshtml using @Html.Raw(@Model.Resources).  This works great for css and js references, but some js components also need to some code to run in the body of the page to initialize.  In those cases, it would be nice to inject those resources scripts into the body.

![image](https://user-images.githubusercontent.com/311063/94875493-ca4a0f00-0409-11eb-8782-f3f8acecc05d.png)

I see two paths that can start a conversation:

A) Add a third resource type to enum ResourceType called BodyScript.  Any resource added as a BodyScript would get injected into the body in the _host.cshtml using @Html.Raw(@Model.BodyScripts). That is a new property that builds <script> tags for each.

B)  Add a ResourceLocation enum with Head and Body members.  Then overload the Resource class to have a ResourceLocation that defaults to Head.  injected resources into the head or body in the _host.cshtml using @Html.Raw(@Model.HeadResources) or into the body using @Html.Raw(@Model.BodyResources)  That is a new property that builds <script> tags for each.

I need this feature for the work I'm doing the IHostResouces and Blazorise.  So I'll likely make a PR for the A case to check out soon.

**Comment by jimspillane** [comment]
[Example #16](https://github.com/muicss/loadjs#documentation) might be used to help implement path B.

**Reply by sbwalker** [reply]
Could you use the JSInterop methods which already exist? Look at themebuilder.razor and you will see that it has JSInterop methods to inject JavaScript for enabling PWA functionality into the body of the page. It uses the following method:

        public Task IncludeScript(string id, string src, string integrity, string crossorigin, string content, string location, string key)

Notice that it already supports location of Head or Body.

**Reply by jimspillane** [reply]
Yes, that is an option.  If there are any dependencies, I would want to test the order of the scripts added; otherwise, we might run into the async timing issues we had previously.

**Reply by sbwalker** [reply]
@jimspillane FYI @markdav-is is referring to using the IHostResources method for injecting these resources ( ie. via _host.cshtml ), so the async timing issues should not be a problem.

**Comment by sbwalker** [comment]
Are you saying that Blazorise does not work when it is registered in the head? 

@markdav-is Please see #764 as there is related conversation there. In regards to the options laid out above my preference would be option B of the items you outlined ( extend Oqtane.Models.Resource class with another property ) - however I think some higher level decisions need to be made before this is introduced.

**Reply by markdav-is** [reply]
Thats right. The code need to run within the body tag so it can find JS references to Window.
